internal/tlscert/tls_cert_validator.go

```diff
@@ -99,67 +105,78 @@ func New(client client.Client) *Validator {
 }
 
 func (v *Validator) Validate(ctx context.Context, config TLSBundle) error {
-	certPEM, err := resolveValue(ctx, v.client, *config.Cert)
+	// Check for missing TLS configuration values
+	err := checkForMissingValues(config)
 	if err != nil {
 		return err
 	}
 
-	keyPEM, err := resolveValue(ctx, v.client, *config.Key)
+	// Resolve values
+	certPEM, keyPEM, caPEM, err := resolveValues(ctx, v.client, config)
 	if err != nil {
 		return err
 	}
 
-	// CA is optional (check if missing)
-	missingCA := config.CA == nil
-
-	var caPEM []byte
-	if !missingCA {
-		caPEM, err = resolveValue(ctx, v.client, *config.CA)
-		if err != nil {
-			return err
-		}
-	}
-
 	// Make the best effort to replace linebreaks in cert/key/ca if present.
 	sanitizedCert := bytes.ReplaceAll(certPEM, []byte("\\n"), []byte("\n"))
 	sanitizedKey := bytes.ReplaceAll(keyPEM, []byte("\\n"), []byte("\n"))
 	sanitizedCA := bytes.ReplaceAll(caPEM, []byte("\\n"), []byte("\n"))
 
-	// Parse the certificate
-	parsedCert, err := parseCertificate(sanitizedCert, ErrCertDecodeFailed, ErrCertParseFailed)
+	// Parse values
+	parsedCert, parsedCAs, err := parseValues(sanitizedCert, sanitizedKey, sanitizedCA)
 	if err != nil {
 		return err
 	}
 
-	// Parse the private key
-	if err := parsePrivateKey(sanitizedKey); err != nil {
+	// Validate values
+	err = validateValues(config, parsedCert, parsedCAs, sanitizedCert, sanitizedKey, v.now())
+	if err != nil {
 		return err
 	}
 
-	// Parse the CA(s)
+	return nil
+}
+
+func checkForMissingValues(config TLSBundle) error {
+	missingCert, missingKey, _ := config.GetMissing()
+
+	if missingCert != missingKey {
+		return ErrMissingCertKeyPair
+	}
+
+	return nil
+}
+
+func parseValues(sanitizedCert, sanitizedKey, sanitizedCA []byte) (*x509.Certificate, []*x509.Certificate, error) {
+	var parsedCert *x509.Certificate
 	var parsedCAs []*x509.Certificate
-	if !missingCA {
-		parsedCAs, err = parseCertificates(sanitizedCA, ErrCADecodeFailed, ErrCAParseFailed)
-		if err != nil {
-			return err
-		}
+	var err error
+
+	// Parse the certificate (if not missing)
+	if sanitizedCert != nil {
+		parsedCert, err = parseCertificate(sanitizedCert, ErrCertDecodeFailed, ErrCertParseFailed)
+	}
+	if err != nil {
+		return parsedCert, parsedCAs, err
 	}
 
-	// Validate certificate
-	if err := validateCertificate(parsedCert, sanitizedCert, sanitizedKey, v.now()); err != nil {
-		return err
+	// Parse the private key (if not missing)
+	if sanitizedKey != nil {
+		err = parsePrivateKey(sanitizedKey)
+	}
+	if err != nil {
+		return parsedCert, parsedCAs, err
 	}
 
-	// Validate CA(s)
-	if !missingCA {
-		for _, ca := range parsedCAs {
-			if err := validateCA(ca, v.now()); err != nil {
-				return err
```

Comment:
I understand the intention of moving all parsing in a separate function, but IMHO it looks more confusing than before:
1. The function is called parse. It takes a key, a cert, and a CA cert and returns 2 certs. I know that we are not interested in the parsed key, but it still looks illogical.
2. It returns 2 certs and it's not obvious from the signature which is the normal cert and which is the CA cert. You could solve it using named return values, but it's also not very common in our codebase.

IMHO, keeping it the way it was is easier to read.